File: src/main/java/net/vidageek/games/regex/task/CaptureGroup.java
package net.vidageek.games.regex.task;

import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import net.vidageek.games.task.JudgedTask;
import net.vidageek.games.task.Task;
import net.vidageek.games.task.status.Ok;

import com.google.common.base.Joiner;
import com.google.common.collect.Lists;

final public class CaptureGroup implements Task {

	private final String matchingTarget;
	private final String[] captureGroupTargets;
	private final List<GroupValidation> validations = Lists.newArrayList();

	public CaptureGroup(final String matchingTarget, final String... captureGroupTargets) {
		this.matchingTarget = matchingTarget;
		this.captureGroupTargets = captureGroupTargets;
		addAllValidations();
	}

	private void addAllValidations() {
		validations.add(new ValidationIfMatch(this.matchingTarget));
		validations.add(new ValidationIfAllCapture(this.captureGroupTargets));
		validations.add(new ValidationCaptureCorrectGroup(this.matchingTarget));
		validations.add(new ValidationIfAllGroupsMatch(this.captureGroupTargets));
	}

	public JudgedTask judge(final String challenge) {
		final Matcher matcher = Pattern.compile(challenge).matcher(matchingTarget);
		return applyAllValidations(challenge, matcher);

	}

	private JudgedTask applyAllValidations(final String challenge, final Matcher matcher) {
		JudgedTask judge = new Ok();
		for (GroupValidation validation : validations) {
			judge = validation.judge(challenge, matcher);
			if (!judge.getOk()) {
				return judge;
			}
		}
		return judge;
	}

	public String getChallenge() {
		return "Qual regex d&aacute; match em " + matchingTarget + " e captura " + captureTarget() + "?";
	}

	private String captureTarget() {
		return Joiner.on(", ").join(captureGroupTargets);
	}

	@Override
	public String toString() {
		return "Capturar " + captureTarget() + " no string " + matchingTarget;
	}

}

File: src/main/scala/net/vidageek/games/regex/task/CaptureGroup.scala
package net.vidageek.games.regex.task;

import scala.collection.mutable._

import net.vidageek.games.regex.task.MatcherTargets._;

import java.util.List
import java.util.regex.Matcher
import java.util.regex.Pattern

import net.vidageek.games.task.JudgedTask
import net.vidageek.games.task.Task
import net.vidageek.games.task.status.Ok

import com.google.common.collect.Lists


class CaptureGroup(val matchingTarget: String, val captureGroupTargets: String*)  extends Task {

	val validations = ListBuffer[GroupValidation]()
	addAllValidations()
	
	def addAllValidations(): Unit = {
		validations += new ValidationIfMatch(matchingTarget)
		validations += new ValidationIfAllCapture(captureGroupTargets.toList : _*)
		validations += new ValidationCaptureCorrectGroup(matchingTarget)
		validations += new ValidationIfAllGroupsMatch(captureGroupTargets.toList : _*)
	}

	def judge(challenge: String): JudgedTask = {
		val matcher: Matcher = Pattern.compile(challenge).matcher(matchingTarget);
		return applyAllValidations(challenge, matcher);

	}

	def applyAllValidations(challenge: String, matcher: Matcher): JudgedTask = {
		validations.foldLeft[JudgedTask](new Ok())((judge, validation) => 
		  	judge.getOk() match {
		  	  case true => validation.judge(challenge, matcher)
		  	  case false => judge 
		  	}
		)
	}

	def getChallenge: String = {
		"Qual regex d&aacute; match em " + from(matchingTarget).asHtml() + " e captura " + captureTarget + "?"
	}

	def captureTarget: String = {
		from(captureGroupTargets.toList : _*).asHtml()
	}

	override def toString(): String = {
		getChallenge
	}

}
